<a href="https://colab.research.google.com/github/CodeHunterOfficial/ML_First_Step/blob/main/%D0%9E%D0%B1%D0%BD%D0%B0%D1%80%D1%83%D0%B6%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B0%D0%BD%D0%BE%D0%BC%D0%B0%D0%BB%D0%B8%D0%B9_%D0%B2_Python_%D1%81_%D0%BF%D0%BE%D0%BC%D0%BE%D1%89%D1%8C%D1%8E_PyOD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## [Обнаружение аномалий с PyOD Python с помощью PyOD](https://towardsdatascience.com/introducing-anomaly-outlier-detection-in-python-with-pyod-40afcccee9ff)

[см. тоже](https://github.com/sayakpaul/FloydHub-Anomaly-Detection-Blog/blob/master/FloydHub%20Anomaly%20Detection%20Blog.ipynb)

## 1. — Установка сцены
В последние годы обнаружение аномалий стало более популярным в сообществе машинного обучения. Несмотря на это, ресурсов на обнаружение аномалий определенно меньше, чем на классические алгоритмы машинного обучения. Таким образом, изучение обнаружения аномалий может показаться более сложным, чем должно быть. Обнаружение аномалий с концептуальной точки зрения на самом деле очень просто!

Наш цель  — дать вам краткое введение в обнаружение аномалий/выбросов. В частности, я покажу вам, как реализовать обнаружение аномалий в Python с помощью пакета PyOD — Python Outlier Detection . Таким образом, вы не только поймете, что такое обнаружение аномалий/выбросов, но и узнаете, как реализовать обнаружение аномалий в Python.

Хорошая новость заключается в том, что PyOD легко применять, особенно если у вас уже есть опыт работы со Scikit-Learn . На самом деле пакет PyOD пытается быть очень похожим на интерфейс API Scikit-Learn.

Предварительные требования: у вас должны быть базовые знания Python и Pandas. Однако никаких знаний по обнаружению аномалий не требуется

## 2. Что такое обнаружение аномалий/выбросов?
Обнаружение аномалий имеет много названий; обнаружение выбросов , анализ выбросов , анализ аномалий и обнаружение новизны . Краткое описание из Википедии описывает обнаружение аномалий следующим образом:
> Обнаружение аномалий — это идентификация редких элементов, событий или наблюдений, которые вызывают подозрения, поскольку значительно отличаются от большинства данных.

Попробуем распаковать приведенные выше утверждения. Скажем, у вас есть набор данных, состоящий из множества наблюдений. Целью обнаружения аномалий является выявление наблюдений, которые значительно отличаются от остальных. Почему вы хотите это сделать? Есть две основные причины:

##Вариант использования 1 — очистка данных
При очистке данных иногда лучше удалить аномалии, поскольку они искажают данные. Давайте проиллюстрируем это на конкретном примере: скажем, вы провели опрос, в котором задаются вопросы о любимых породах кошек респондентов 

Сначала вы даете опрос 100 людям, каждый из которых проходит опрос. Теперь вы хотите оценить среднее время, затрачиваемое на прохождение опроса. Почему? Вы хотите, чтобы в опросе приняли участие еще 10 000 человек. Было бы профессионально указать примерно, сколько времени занимает опрос новых респондентов. Несмотря на то, что кошки потрясающие, люди заняты!

Допустим, вы получили следующие результаты от первых 100 человек:

* 3 минуты — 57 респондентов
* 4 минуты — 33 респондента
* 5 минут — 6 респондентов
* 6 минут — 3 респондента
* 480 минут — 1 респондент

Что происходит с последним? 480 минут это 8 часов! При дальнейшем осмотре вы обнаружите, что респондент начал опрос в 23:58 вечера, а затем стоял неподвижно с 00:00 до 07:56. Потом с 07:56 до 07:58 закончилось. Вы видите, что произошло?

Ясно, что человек начал опрос, затем лег спать, а затем закончил опрос, проснувшись утром. Если вы сохраните этот результат, то среднее время прохождения опроса будет

среднее = (3 * 57 + 4 * 33 + 5 * 6 + 6 * 3 + 1 * 480)/100 = 8,31

Однако утверждение, что опрос занимает примерно 8 минут, неверно. Единственная причина, по которой это заняло так много времени, это из-за сонливого респондента 

Правильнее было бы убрать этого человека из подсчета и получить

среднее = (3 * 57 + 4 * 33 + 5 * 6 + 6 * 3)/99 = 3,54

Для простоты опрос можно записать в виде предложения: Среднее время завершения опроса составляет от 3 до 4 минут.

Здесь вы вручную удалили выброс, чтобы очистить данные, чтобы лучше представить реальность. Обнаружение аномалий реализует алгоритмы для автоматического обнаружения выбросов.

> Предупреждение: в приведенном выше примере вы удалили выброс, чтобы лучше сопоставить длину опроса с реальностью. Обнаружение аномалий никогда не должно использоваться для искусственного создания лучшего продукта, чем он есть на самом деле. Следует внимательно рассмотреть вопрос об этичности использования обнаружения аномалий для очистки данных.

Вариант использования 2 — Прогноз
В других приложениях интерес представляют сами аномалии . Примерами являются вторжение в сеть, банковское мошенничество и некоторые структурные дефекты. В этих приложениях аномалии представляют собой то, что заслуживает дальнейшего изучения.

* Вторжение в сеть — аномалии в сетевых данных могут указывать на то, что произошла какая-то сетевая атака.
* Банковское мошенничество — аномалии в данных о транзакциях могут указывать на мошенничество или подозрительное поведение.
* Структурные дефекты — аномалии могут указывать на то, что с вашим оборудованием что-то не так. Хотя в этом параметре обычно доступно более традиционное программное обеспечение для мониторинга, обнаружение аномалий может обнаружить более странные дефекты.

>Предостережение: в некоторых случаях, например, в случае банковского мошенничества, подозрения не всегда вызывает отдельная транзакция. Следует учитывать частоту и величину множественных транзакций, рассматриваемых в контексте. Чтобы справиться с этим, данные должны быть агрегированы соответствующим образом. Это выходит за рамки этого блога, но вам следует об этом знать.

##3 — Знакомство с PyOD
Давайте опишем пакет Python PyOD , который помогает вам обнаруживать аномалии. По словам документации PyOD:

>PyOD — это всеобъемлющий и масштабируемый набор инструментов Python для обнаружения посторонних объектов в многомерных данных.

Короче говоря, PyOD предоставляет вам набор моделей, которые выполняют обнаружение аномалий. Некоторые интересные моменты, которые стоит упомянуть:

>PyOD включает в себя более 30 различных алгоритмов. Некоторые из них — классика (например, LOF), а другие — новинки (например, COPOD). [См. полный список поддерживаемых моделей](https://pyod.readthedocs.io/en/latest/pyod.models.html).

>PyOD оптимизировал свой код с помощью jit-decorator от Numba. [См. в блоге о Numba](https://towardsdatascience.com/how-to-skyrocket-your-python-speed-with-numba-89fd16362e47), если вам это интересно.

>PyOD имеет единый API. Следовательно, если вы познакомитесь с несколькими моделями в PyOD, то сможете легко изучить остальные. Я рекомендую взглянуть на [PyOD API CheatSheet](https://pyod.readthedocs.io/en/latest/api_cc.html) после того, как вы закончите этот блог.

Если вы используете PIP, то вы можете установить PyOD с помощью команды:

In [17]:
!pip install pyod

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Если у вас уже был установлен PyOD ранее, убедитесь, что он обновлен с помощью команды pip:

In [38]:
!pip install --upgrade pyod

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


Если вместо этого вы используете диспетчер пакетов Conda, вы можете запустить команду:

In [39]:
!conda install -c conda-forge pyod

/bin/bash: conda: command not found


В этом сообщении блога я продемонстрирую два алгоритма обнаружения аномалий: KNN и LOC. Возможно, вы уже слышали о KNN (K — ближайшие соседи) ранее, тогда как LOC (локальный посторонний фактор), вероятно, вам не знаком. Давайте сначала посмотрим на данные, которые вы будете использовать.

## 4 — Знакомство с данными
Мы будем использовать классический набор данных Titanic. Чтобы загрузить набор данных в Pandas, просто запустите код ниже:

In [18]:
import pandas as pd
titanic = pd.read_csv(
"https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
)

In [19]:
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

Чтобы проверить первые строки набора данных, используйте head()метод:

In [20]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Как видите, есть столбцы, представляющие пол, возраст, стоимость проезда, пассажирский класс, билет и т. д. Для простоты вы будете работать только со следующими четырьмя столбцами:

In [21]:
# Selecting only the columns Survived, Pclass, Fare, and Sex
partial_titanic = titanic[["Survived", "Pclass", "Fare", "Sex"]]

In [22]:
partial_titanic 

Survived  Pclass     Fare     Sex
0           0       3   7.2500    male
1           1       1  71.2833  female
2           1       3   7.9250  female
3           1       1  53.1000  female
4           0       3   8.0500    male
..        ...     ...      ...     ...
886         0       2  13.0000    male
887         1       1  30.0000  female
888         0       3  23.4500  female
889         1       1  30.0000    male
890         0       3   7.7500    male

[891 rows x 4 columns]

В нем нет пропущенных значений partial_titanic. Однако столбец Sexсостоит из строковых значений maleили female. Чтобы иметь возможность обнаруживать аномалии, вам нужны числовые значения. Вы можете преобразовать эту двоичную категориальную переменную в значения 0 и 1 с помощью кода:

In [23]:
# Change the categorical value Sex to numeric values
partial_titanic["Sex"] = partial_titanic["Sex"].map({
"male": 0,
"female": 1
})

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [24]:
partial_titanic["Sex"]

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    1
889    0
890    0
Name: Sex, Length: 891, dtype: int64

Теперь вы готовы заняться обнаружением аномалий:

## 5 — Обнаружение аномалий для очистки данных

Давайте теперь воспользуемся обнаружением аномалий, чтобы очистить набор данных partial_titanic, который вы создали в предыдущем разделе. Для этого вы будете использовать модель KNN. Модель KNN исследует данные и ищет точки данных (строки), которые находятся далеко от других точек данных. Для начала вы импортируете модель KNN следующим образом:

In [25]:
# Import the KNN
from pyod.models.knn import KNN

Начнем с запуска модели KNN:

In [26]:
# Initiate a KNN model
KNN_model = KNN()

Для методов обнаружения аномалий для очистки данных вы можете использовать весь набор данных следующим образом.

In [27]:
# Fit the model to the whole dataset
KNN_model.fit(partial_titanic)

KNN(algorithm='auto', contamination=0.1, leaf_size=30, method='largest',
  metric='minkowski', metric_params=None, n_jobs=1, n_neighbors=5, p=2,
  radius=1.0)

При запуске приведенного выше кода вы получаете множество значений по умолчанию (например, contamination=0.1). Это можно настроить при необходимости. После запуска модели вы можете получить доступ к двум типам вывода:

* **Метки:** запустив KNN_model.labels_, вы можете найти двоичные метки того, является ли наблюдение выбросом или нет. Число 0 указывает на нормальное наблюдение, а число 1указывает на выброс.
* **Оценки принятия решений**: запустив KNN_model.decision_scores_, вы получаете сырые оценки того, насколько что-то является выбросом. Значения будут варьироваться от 0и вверх. Более высокий показатель аномалии указывает на то, что точка данных является скорее выбросом.

Давайте проверим метки обученной модели:

In [28]:
# Find the labels
outlier_labels = KNN_model.labels_
# Find the number of outliers
number_of_outliers = len(outlier_labels[outlier_labels == 1])
print(number_of_outliers)

88


Для набора данных с 891 пассажиром 88 выбросов — это довольно много. Чтобы уменьшить это, вы можете указать параметр contaminationв модели KNN, чтобы он был ниже. Указывает contaminationпроцент точек данных, которые являются выбросами. Предположим, что загрязнение составляет только 1%:

In [29]:
# Initiate a KNN model
KNN_model = KNN(contamination=0.01)
# Fit the model to the whole dataset
KNN_model.fit(partial_titanic)
# Find the labels
outlier_labels = KNN_model.labels_
# Find the number of outliers
number_of_outliers = len(outlier_labels[outlier_labels == 1])
print(number_of_outliers)

9


Теперь только 9аутсайдеры! Вы можете проверить их:

In [30]:
# Finding the outlier passengers
outliers = partial_titanic.iloc[outlier_labels == 1]

Если вы проверите outliersпеременную, вы получите следующую таблицу:

In [31]:
outliers

Survived  Pclass      Fare  Sex
258         1       1  512.3292    1
380         1       1  227.5250    1
679         1       1  512.3292    0
689         1       1  211.3375    1
700         1       1  227.5250    1
716         1       1  227.5250    1
730         1       1  211.3375    1
737         1       1  512.3292    0
779         1       1  211.3375    1

Если вы посмотрите на пассажиров выше, то модель KNN увидит, что стоимость их проезда невероятно высока. Среднюю стоимость проезда для всех пассажиров можно легко найти в Pandas:

In [33]:
# Average fare price
round(partial_titanic["Fare"].mean(), 3)

32.204

Алгоритм KNN успешно нашел 9 пассажиров, которые являются выбросами в смысле стоимости проезда. Есть [много дополнительных параметров](https://pyod.readthedocs.io/en/latest/pyod.models.html#module-pyod.models.knn), с которыми вы можете поиграться для модели KNN, чтобы она соответствовала вашим конкретным потребностям 

Выбросы теперь можно удалить из данных, если вы чувствуете, что они не отражают общее впечатление от данных. Как упоминалось ранее, вам следует тщательно обдумать, подходит ли обнаружение аномалий для очистки данных для вашей проблемы.

### 6 — Обнаружение аномалий для прогнозирования

В предыдущем разделе вы рассмотрели обнаружение аномалий для очистки данных . В этом разделе вы возьмете пик обнаружения аномалий для предсказания . Вы будете обучать модель на существующих данных, а затем использовать модель, чтобы предсказать, являются ли новые данные выбросами.

Скажем, распространился слух, что миссис Уотсон также забрала «Титаник», но ее смерть так и не была зарегистрирована. По слухам, миссис Уотсон была богатой дамой, которая заплатила 1000 долларов за путешествие на «Титанике» в очень эксклюзивном номере.

Обнаружение аномалии не может с уверенностью сказать, является ли слух правдивым или ложным. Однако он может сказать, является ли миссис Ватсон аномалией или нет, основываясь на информации других пассажиров. Если она аномалия, слух следует воспринимать с долей скептицизма.

Давайте проверим существование миссис Ватсон с другой моделью в библиотеке PyOD; **Фактор локального выброса (LOF)** . Модель LOF проверяет, является ли точка данных выбросом, сравнивая локальную плотность точки данных с локальными плотностями ее соседей. Для получения дополнительной информации об этом методе вы можете посетить [его страницу в Википедии](https://en.wikipedia.org/wiki/Local_outlier_factor).

Приступаем к кодированию! Вы начинаете с создания модели Local Outlier Factor:

In [34]:
# Import the LOF
from pyod.models.lof import LOF
# Initiate a LOF model
LOF_model = LOF()
# Train the model on the Titanic data
LOF_model.fit(partial_titanic)

LOF(algorithm='auto', contamination=0.1, leaf_size=30, metric='minkowski',
  metric_params=None, n_jobs=1, n_neighbors=20, novelty=True, p=2)

Обратите внимание, насколько работа с моделью LOF похожа на работу с моделью [KNN](https://russianblogs.com/article/1087498396/).

Теперь вы можете представить миссис Ватсон как точку данных:

In [35]:
# Represent Mrs. Watson as a data point(Представьте миссис Ватсон в виде точки данных)
mrs_watson = [[0, 1, 1000, 1]]

In [36]:
mrs_watson

[[0, 1, 1000, 1]]

Значения mrs_watsonпредставляют ее выживание ( 0для не выживших), пассажирский класс ( 1для первого класса), стоимость проезда ( 1000$для цены проезда) и пол ( 1для женщин). Для модели LOF требуются двумерные массивы, поэтому []в mrs_watson.

Теперь мы используем predict()метод, чтобы предсказать, является ли миссис Ватсон выбросом или нет:

In [37]:
outlier = LOF_model.predict(mrs_watson)
print(outlier)

[1]


Значение 1 указывает на то, что миссис Ватсон является исключением. Это должно вызвать у вас подозрение, что слухи о миссис Уотсон ложны

## 7 — Подведение итогов
Я показал вам, как реализовать обнаружение аномалий с помощью двух алгоритмов KNN и LOF. Как вы, вероятно, подозреваете, в PyOD есть еще [много алгоритмов, с которыми вы можете поиграться](https://pyod.readthedocs.io/en/latest/pyod.html).

Обнаружение аномалий важно как для очистки данных, так и для прогнозирования выбросов. Приложение под рукой должно определить, представляет ли интерес применение обнаружения аномалий. Если вы планируете применять обнаружение аномалий в Python, то [PyOD](https://pyod.readthedocs.io/en/latest/) — хороший выбор.